File: BotNet/client/CrazyUncleBurton/java/Virus.java
import com.sun.jna.Pointer;

///Stores the information about a virus
class Virus extends Mappable 
{
  public Virus(Pointer p)
  {
    super(p);
  }
  boolean validify()
  {
    if(iteration == BaseAI.iteration) return true;
    for(int i = 0; i < BaseAI.viruses.length; i++)
    {
      if(BaseAI.viruses[i].ID == ID)
      {
        ptr = BaseAI.viruses[i].ptr;
        iteration = BaseAI.iteration;
        return true;
      }
    }
    throw new ExistentialError();
  }
    
    //commands
    
  ///Moves a unit to the given (x,y) position
  int move(int x, int y)
  {
    validify();
    return Client.INSTANCE.virusMove(ptr, x, y);
  }
    
    //getters
    
  ///Unique Identifier
  public int getId()
  {
    validify();
    return Client.INSTANCE.virusGetId(ptr);
  }
  ///The x coordinate
  public int getX()
  {
    validify();
    return Client.INSTANCE.virusGetX(ptr);
  }
  ///The y cooridnate
  public int getY()
  {
    validify();
    return Client.INSTANCE.virusGetY(ptr);
  }
  ///The owner of this Virus
  public int getOwner()
  {
    validify();
    return Client.INSTANCE.virusGetOwner(ptr);
  }
  ///The Virus's level
  public int getLevel()
  {
    validify();
    return Client.INSTANCE.virusGetLevel(ptr);
  }
  ///The number of times this virus can still move this turn
  public int getMovesLeft()
  {
    validify();
    return Client.INSTANCE.virusGetMovesLeft(ptr);
  }
  ///This virus is alive if the function returns a 1
  public int getLiving()
  {
    validify();
    return Client.INSTANCE.virusGetLiving(ptr);
  }

}

File: BotNet/client/CrazyUncleBurton/c/Virus.cpp
// -*-c++-*-

#include "Virus.h"
#include "game.h"


Virus::Virus(_Virus* pointer)
{
    ptr = (void*) pointer;
}

int Virus::id()
{
  return ((_Virus*)ptr)->id;
}

int Virus::x()
{
  return ((_Virus*)ptr)->x;
}

int Virus::y()
{
  return ((_Virus*)ptr)->y;
}

int Virus::owner()
{
  return ((_Virus*)ptr)->owner;
}

int Virus::level()
{
  return ((_Virus*)ptr)->level;
}

int Virus::movesLeft()
{
  return ((_Virus*)ptr)->movesLeft;
}

int Virus::living()
{
  return ((_Virus*)ptr)->living;
}


int Virus::move(int x, int y)
{
  return virusMove( (_Virus*)ptr, x, y);
}



std::ostream& operator<<(std::ostream& stream,Virus ob)
{
  stream << "[Virus" 
         << " id(" << ((_Virus*)ob.ptr)->id  << ')'
         << " (" << ((_Virus*)ob.ptr)->x  << ',' << ((_Virus*)ob.ptr)->y << ')'
         << " owner(" << ((_Virus*)ob.ptr)->owner  << ')'
         << " Lvl( " << ((_Virus*)ob.ptr)->level  << ')'
         << " movesLeft(" << ((_Virus*)ob.ptr)->movesLeft  << ')'
         << " living(" << ((_Virus*)ob.ptr)->living  << ')'
         << "]"
    ;
  return stream;
}
